### PR TITLE
Fairness evaluations

### PR DESCRIPTION
**Pull request by aloukina**
This PR implements fairness evaluations proposed in https://aclweb.org/anthology/papers/W/W19/W19-4401/:

- Added  new code/notebook
- Updated code to include new notebook
- Updated documentation
- Updated existing tests where necessary

Note: the new files added to test outputs are in #267  for ease of reviewing.

### CODE REVIEW

**Comment by desilinguist** [comment]
There's a bug in `config_rsmeval.rst` where one of the section headers is using `=` and all of other headers use `~`. This makes all of the optional sections not show up. I am going to fix this with a commit.

**Comment by desilinguist** [comment]
I feel like the fairness analyses are a big enough deal that we should somehow point them out more prominently rather than just burying them in nested menus where they don't even show up in the side bar.

File: doc/config_rsmeval.rst [hunk 1]
```diff
@@ -132,6 +132,15 @@ RSMTool provides pre-defined sections for ``rsmeval`` (listed below) and, by def
 
     - ``evaluation by group``: Shows barplots with the main evaluation metrics by each of the subgroups specified in the configuration file.
 
+
+    - ``fairness_analyses``: Additional fairness analyses suggested in `Loukina, Madnani, & Zechner, 2019 <https://aclweb.org/anthology/papers/W/W19/W19-4401/>`_. The notebook shows:
+
+        - percentage of variance in squared error explained by subgroup membership
+        - percentage of variance in absolute error explained by subgroup membership
```
Comment (desilinguist) [inline doc/config_rsmeval.rst]: So, we aren't using the signed difference like we said in the paper?

File: doc/intermediate_files_rsmeval.rst [hunk 2]
```diff
@@ -131,4 +131,17 @@ Evaluations based on test theory
...
\ No newline at end of file
+- ``true_score_eval``: evaluations of system scores against estimated true score. Contains total counts of single and double-scored response, variances for human and system scores for these sets of responses, and mean squared error (MSE) and proportional reduction in mean squared error (PRMSE) when predicting true score using system score. 
+
+Additional fairness analyses
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These files contain the results of additional fairness analyses suggested in suggested in `Loukina, Madnani, & Zechner, 2019 <https://aclweb.org/anthology/papers/W/W19/W19-4401/>`_. 
+
+- ``<METRICS>_by_<SUBGROUP>.ols``: a serialized object of type ``pandas.stats.ols.OLS`` containing the fitted model for estimating the variance attributed to a given subgroup membership for a given metrics. The subgroups are defined by the :ref:`configuration file<subgroups_eval>`. The metrics are ``osa`` (overall score accuracy), ``osd`` (overall score difference), and ``csd`` (conditional score difference).
```
Comment (desilinguist) [inline doc/intermediate_files_rsmeval.rst]:
```suggestion
- ``<METRICS>_by_<SUBGROUP>.ols``: a serialized object of type ``pandas.stats.ols.OLS`` containing the fitted model for estimating the variance attributed to a given subgroup membership for a given metric. The subgroups are defined by the :ref:`configuration file<subgroups_eval>`. The metrics are ``osa`` (overall score accuracy), ``osd`` (overall score difference), and ``csd`` (conditional score difference). 
```

File: doc/intermediate_files_rsmtool.rst [hunk 3]
```diff
@@ -245,4 +245,17 @@ Evaluations based on test theory
...
\ No newline at end of file
+- ``true_score_eval``: evaluations of system scores against estimated true score. Contains total counts of single and double-scored response, variances for human and system scores for these sets of responses, and mean squared error (MSE) and proportional reduction in mean squared error (PRMSE) when predicting true score using system score. 
+
+Additional fairness analyses
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+These files contain the results of additional fairness analyses suggested in suggested in `Loukina, Madnani, & Zechner, 2019 <https://aclweb.org/anthology/papers/W/W19/W19-4401/>`_. 
+
+- ``<METRICS>_by_<SUBGROUP>.ols``: a serialized object of type ``pandas.stats.ols.OLS`` containing the fitted model for estimating the variance attributed to a given subgroup membership for a given metrics. The subgroups are defined by the :ref:`configuration file<subgroups_rsmtool>`. The metrics are ``osa`` (overall score accuracy), ``osd`` (overall score difference), and ``csd`` (conditional score difference).
```
Comment (desilinguist) [inline doc/intermediate_files_rsmtool.rst]:
```suggestion
- ``<METRICS>_by_<SUBGROUP>.ols``: a serialized object of type ``pandas.stats.ols.OLS`` containing the fitted model for estimating the variance attributed to a given subgroup membership for a given metric. The subgroups are defined by the :ref:`configuration file<subgroups_rsmtool>`. The metrics are ``osa`` (overall score accuracy), ``osd`` (overall score difference), and ``csd`` (conditional score difference). 
```

File: rsmtool/fairness_utils.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+import statsmodels.formula.api as smf
+from statsmodels.stats.anova import anova_lm
+
+from rsmtool.writer import DataWriter
+from rsmtool.container import DataContainer
+
+
+
+def convert_to_ordered_category(group_values):
+    """Convert supplied series to an ordered category
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
    """
    Convert supplied series to an ordered category
```

File: rsmtool/fairness_utils.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                                                                   ordered=True)
+    return group_ordered_category
+    
+
+
+
+def get_coefficients(fit,
+                     base_category):
+    
+    """ Extract estimates, significance and confidence intervals
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
    """
    Extract estimates, significance and confidence intervals
```

File: rsmtool/fairness_utils.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    for members of the group of interest. The names of the 
+    predictors are processed to remove the prefix added by Statmodels.
+    The name of the base category is added in parenthesis to the Intercept. 
+
+    Parameters:
+    -----------
+    fit: fit
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]: Should this be `pandas.stats.ols.OLS`?

File: rsmtool/fairness_utils.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        Linear regression object fitted using statsmodels
+
+    base_category: str
+        Name of the group used a reference category when fitting the model
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
        Name of the group used as reference category when fitting the model
```

File: rsmtool/fairness_utils.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    group_category = group_values.astype("category")
+    group_ordered_category = group_category.cat.reorder_categories(groups_by_size,
+                                                                   ordered=True)
+    return group_ordered_category
+    
+
+
+
+def get_coefficients(fit,
+                     base_category):
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]: I don't think you need to have these on separate lines.

File: rsmtool/fairness_utils.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+                           output_dir,
+                           experiment_id, 
+                           file_format):
+    """
+    Save the results of fairness analysis to disk.
+
+    Parameters
+    ----------
+    fit_dictionary: dict
+        A dictionary of fitted models generated by get_fairness_analysis
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
        A dictionary of fitted models generated by `get_fairness_analysis()`
```

File: rsmtool/fairness_utils.py [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    fairness_container: container.DataContainer
+        A data container with the results of fairness analysis generated by get_fairness_analysis
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
        A data container with the results of fairness analysis generated by `get_fairness_analysis()`
```

File: rsmtool/fairness_utils.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    # Now let's write out the content of the data container
+    writer = DataWriter(experiment_id)
+    writer.write_experiment_output(output_dir,
+                                   fairness_container,
+                                   file_format=file_format,
+                                   index=True)
+
+    
+    
+def get_fairness_analysis(df,
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
def get_fairness_analyses(df,
```

File: rsmtool/notebooks/fairness_analyses.ipynb [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    "        imgfile = join(figure_dir, '{}_fairness_estimates_{}.svg'.format(experiment_id, group))\n",
+    "        plt.savefig(imgfile)\n",
+    "        if use_thumbnails:\n",
+    "            show_thumbnail(imgfile, next(id_generator))\n",
+    "        else:\n",
+    "            plt.show()\n",
+    "    \n",
+    "    \n",
+    "    # Show the conditional score plot\n",
+    "    display(Markdown(\"The plot shows average {} system score for each group conditioned \"\n",
```
Comment (desilinguist) [inline rsmtool/notebooks/fairness_analyses.ipynb]:
```suggestion
    "    display(Markdown(\"The plot shows average `{}` system score for each group conditioned \"\n",
```

File: rsmtool/notebooks/fairness_analyses.ipynb [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    "                     \"on human score\".format(system_score_column)))\n",
```
Comment (desilinguist) [inline rsmtool/notebooks/fairness_analyses.ipynb]:
```suggestion
    "                     \"on the human score.\".format(system_score_column)))\n",
```

File: doc/config_rsmeval.rst (hunk 1 above)
Comment (aloukina) [reply doc/config_rsmeval.rst]: API: done in the evaluation PR.

Comment (aloukina) [reply doc/config_rsmeval.rst]: Absolute: Oops. That's what it was supposed to be. I used absolute in a wrong sense. Will fix now.

File: rsmtool/fairness_utils.py (hunk 11 above)
Comment (aloukina) [reply rsmtool/fairness_utils.py]: Will also change the function calls.

File: rsmtool/fairness_utils.py [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+                          group,
+                          system_score_column,
+                          human_score_column='sc1'):
+    """ 
+    Compute fairness analyses described in Loukina et al. (2019) (https://aclweb.org/anthology/papers/W/W19/W19-4401/)
```
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/fairness_utils.py]: Will do in the evaluation branch where I can make sure it works if that's ok.

File: tests/test_fairness_utils.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+                              'Esperanto',
+                              'Klingon',
+                              'Esperanto']})
+    (model_dict,
+     dc) = get_fairness_analysis(df,
+                                'L1',
+                                 'raw')
+    eq_(len(model_dict), 3)
+    eq_(len(dc), 4)
+    print(dc['fairness_metrics_by_L1'])
```
Comment (jbiggsets) [inline tests/test_fairness_utils.py]: Do we need this print statement?

File: rsmtool/notebooks/fairness_analyses.ipynb [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+    "                                                                     float_format=float_format_func)))\n",
+    "    # show coefficient plots\n",
+    "    # define groups and color palette\n",
+    "    groups_by_size = df_pred_preproc_merged[group].value_counts().index\n",
+    "    colors = sns.color_palette(\"Greys_r\", len(groups_by_size))\n",
+    "\n",
+    "    \n",
+    "    display(Markdown(\"The plots show error estimates for different categories for each group \"\n",
+    "                     \"(squared error for OSA, absolute error for OSD, and conditional error for CSD) \"\n",
+    "                     \"The estimates have been adjusted for Intercept\"\n",
```
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/notebooks/fairness_analyses.ipynb]: based -> used. Ok?

File: tests/test_fairness_utils.py (hunk 15 above)
Comment (aloukina) [reply tests/test_fairness_utils.py]: Oops! Of course not, thank you!

File: rsmtool/fairness_utils.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/fairness_utils.py]: Done in the other branch.

File: rsmtool/notebooks/fairness_analyses.ipynb (hunk 16 above)
[the remark this answers is not in the document]
Comment (aloukina) [reply rsmtool/notebooks/fairness_analyses.ipynb]: Fixed.

File: rsmtool/fairness_utils.py [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+                           file_format):
+    """
+    Save the results of fairness analysis to disk.
+
+    Parameters
+    ----------
+    fit_dictionary: dict
+        A dictionary of fitted models generated by `get_fairness_analysis()`
+    fairness_container: container.DataContainer
+        A data container with the results of fairness analysis generated by `get_fairness_analysis()`
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
        A data container with the results of fairness analysis generated by `get_fairness_analyses()`
```

File: rsmtool/fairness_utils.py [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+                           output_dir,
+                           experiment_id, 
+                           file_format):
+    """
+    Save the results of fairness analysis to disk.
+
+    Parameters
+    ----------
+    fit_dictionary: dict
+        A dictionary of fitted models generated by `get_fairness_analysis()`
```
Comment (desilinguist) [inline rsmtool/fairness_utils.py]:
```suggestion
        A dictionary of fitted models generated by `get_fairness_analyses()`
```

File: rsmtool/notebooks/fairness_analyses.ipynb [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+    "                                                                     float_format=float_format_func)))\n",
+    "    # show coefficient plots\n",
+    "    # define groups and color palette\n",
+    "    groups_by_size = df_pred_preproc_merged[group].value_counts().index\n",
+    "    colors = sns.color_palette(\"Greys_r\", len(groups_by_size))\n",
+    "\n",
+    "    \n",
+    "    display(Markdown(\"The plots show error estimates for different categories for each group \"\n",
+    "                     \"(squared error for OSA, absolute error for OSD, and conditional error for CSD). \"\n",
+    "                     \"The estimates have been adjusted for the value of the group used as the Intercept\"\n",
```
Comment (desilinguist) [inline rsmtool/notebooks/fairness_analyses.ipynb]:
```suggestion
    "                     \"The estimates have been adjusted for the value of the group used as the Intercept.\"\n",
```

File: rsmtool/fairness_utils.py (hunk 17 above)
Comment (aloukina) [reply rsmtool/fairness_utils.py]: Oh, I fixed it in another branch. Let me fix it here too :)

**Pull request merged**